File: 133.Clone Graph/Solution.py
# Definition for a undirected graph node
# class UndirectedGraphNode(object):
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class Solution(object):
    def cloneGraph(self, node):
        """
        :type node: UndirectedGraphNode
        :rtype: UndirectedGraphNode
        """
        if not node:
            return
        root = UndirectedGraphNode(node.label)
        stack = [node]
        visit = {}
        visit[node.label] = root
        while stack:
            top = stack.pop()
            for n in top.neighbors:
                if n.label not in visit:
                    stack.append(n)
                    visit[n.label] = UndirectedGraphNode(n.label)
                visit[top.label].neighbors.append(visit[n.label])
        return root

        # if not node:
        #     return
        # root = UndirectedGraphNode(node.label)
        # stack = [node]
        # visit = []
        # visit.append(node)
        # while stack:
        #     top = stack.pop()
        #     for n in top.neighbors:
        #         temp = 0
        #         for k in visit:
        #             if n.label == k.label:
        #                 top.neighbors.append(k)
        #                 break
        #             else:
        #                 temp += 1
        #         if temp == len(visit):
        #             stack.append(n)
        #             new = UndirectedGraphNode(n.label)
        #             visit.append(new)
        #             top.neighbors.append(new)
        # return root

File: 133.Clone Graph/src/Solution.java
import java.util.HashMap;
import java.util.Map;

/**
 * Definition for undirected graph.
 * class UndirectedGraphNode {
 *     int label;
 *     List<UndirectedGraphNode> neighbors;
 *     UndirectedGraphNode(int x) { label = x; neighbors = new ArrayList<UndirectedGraphNode>(); }
 * };
 */
public class Solution {
	Map<Integer, UndirectedGraphNode> nodes = new HashMap<Integer, UndirectedGraphNode>();
    public UndirectedGraphNode cloneGraph(UndirectedGraphNode node) {
        if (node == null)
        	return null;
        UndirectedGraphNode clone = null;
        if (!nodes.containsKey(node.label)) {
        	clone = new UndirectedGraphNode(node.label);
        	nodes.put(clone.label, clone);
        }
        else{
        	return nodes.get(node.label);
        }
        for (int i = 0; i < node.neighbors.size(); i++) {
        	if (node.neighbors.get(i).label == node.label) {
        		clone.neighbors.add(clone);
        	}
        	else{
        		clone.neighbors.add(cloneGraph(node.neighbors.get(i)));
        	}
        }
        return clone;
        
    }
}